torch/csrc/jit/frontend/exit_transforms.cpp

```diff
@@ -303,6 +303,168 @@ struct ExitTransformer {
     return ExitPair(has_exited, exit_vals);
   }
 
+  // Adds prim::With nodes to a block to help handle early exits between
+  // prim::Enter and prim::Exit nodes. More specifically, it transforms
+  // IR that looks like this:
+  //
+  //   %a = prim::Enter(%b)
+  //   <code>
+  //   %c = prim::Exit(%b)
+  //
+  // to this:
+  //
+  //   %a = prim::Enter(%b)
+  //   = prim::With()
+  //     block0():
+  //       <code>
+  //     -> ()
+  //     block1():
+  //       %c = prim::Exit(%b)
+  //     -> ()
+  //
+  void addWithNodes(Block* block) {
+    // First, find all Enter-Exit pairs up front to avoid iterator invalidation
+    // issues later when moving nodes around. Do this by iterating through the
+    // nodes of the block while keeping a stack of encountered Enter nodes. Each
+    // time an Exit node is seen, its corresponding Enter node must be at the
+    // top of the stack. Pop it and record the pair.
+    std::vector<std::pair<Node*, Node*>> enter_exit_pairs;
+    std::vector<Node*> stack;
+
+    for (auto it = block->nodes().begin(); it != block->nodes().end(); ++it) {
+      if (it->kind() == prim::Enter && it->next()->kind() != prim::With) {
```

Comment:
It was needed because I was doing the transformation to/from the block-based representation in `transformExits`, which gets called recursively on the body block of a `prim::With` and might end up processing `prim::With`. But it's no longer needed because I moved the transformation of the `Enter/Exit` pairs to and from `With` nodes to the location you suggested so that we only do those transforms once.